op-proposer/proposer/driver.go

```diff
@@ -309,33 +389,56 @@ func (l *L2OutputSubmitter) loop() {
 
 	ctx := l.ctx
 
-	ticker := time.NewTicker(l.Cfg.PollInterval)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ticker.C:
-			output, shouldPropose, err := l.FetchNextOutputInfo(ctx)
-			if err != nil {
-				break
+	if l.dgfContract == nil {
+		ticker := time.NewTicker(l.Cfg.PollInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				output, shouldPropose, err := l.FetchNextOutputInfo(ctx)
+				if err != nil || !shouldPropose {
+					break
+				}
+
+				l.proposeOutput(ctx, output)
+			case <-l.done:
+				return
 			}
-			if !shouldPropose {
-				break
-			}
-			cCtx, cancel := context.WithTimeout(ctx, 10*time.Minute)
-			if err := l.sendTransaction(cCtx, output); err != nil {
-				l.Log.Error("Failed to send proposal transaction",
-					"err", err,
-					"l1blocknum", output.Status.CurrentL1.Number,
-					"l1blockhash", output.Status.CurrentL1.Hash,
-					"l1head", output.Status.HeadL1.Number)
-				cancel()
-				break
+		}
+	} else {
+		ticker := time.NewTicker(l.Cfg.ProposalInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				blockNumber, err := l.FetchCurrentBlockNumber(ctx)
+				if err != nil {
+					break
+				}
+
+				output, shouldPropose, err := l.fetchOutput(ctx, blockNumber)
+				if err != nil || !shouldPropose {
+					break
+				}
+
+				l.proposeOutput(ctx, output)
+			case <-l.done:
+				return
 			}
-			l.Metr.RecordL2BlocksProposed(output.BlockRef)
-			cancel()
-
-		case <-l.done:
-			return
 		}
 	}
 }
+
+func (l *L2OutputSubmitter) proposeOutput(ctx context.Context, output *eth.OutputResponse) {
+	cCtx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	if err := l.sendTransaction(cCtx, output); err != nil {
+		l.Log.Error("Failed to send proposal transaction",
+			"err", err,
+			"l1blocknum", output.Status.CurrentL1.Number,
+			"l1blockhash", output.Status.CurrentL1.Hash,
+			"l1head", output.Status.HeadL1.Number)
+		cancel()
+	}
+	l.Metr.RecordL2BlocksProposed(output.BlockRef)
+	cancel()
```

Comment:
Now that this part is nicely extracted, it's better to `defer cancel()`
```suggestion
	cCtx, cancel := context.WithTimeout(ctx, 10*time.Minute)
	defer cancel()
	if err := l.sendTransaction(cCtx, output); err != nil {
		l.Log.Error("Failed to send proposal transaction",
			"err", err,
			"l1blocknum", output.Status.CurrentL1.Number,
			"l1blockhash", output.Status.CurrentL1.Hash,
			"l1head", output.Status.HeadL1.Number)
		return // return here on error? otherwise below metrics recording might be misleading
	}
	l.Metr.RecordL2BlocksProposed(output.BlockRef)
```
and there might be a missing `return` in the error path

Reply:
I also realize that the behavior changed. Before, on an error, there was a `break` of the loop. Maybe you want to return the `err` from `proposeOutput` and break on an error in the loop?

Reply:
Good call on the defer, don't believe the `break` broke out of the loop, just the switch. Updated!